Globus Flow: Tomography Reconstruction at ALCF

davramov opened the pull request:
# Test environment
I isolated the file tree for bl832 and dependencies, specifically this Globus reconstruction flow.
### File tree

    ALCF_compute_test
    │   ├── config.yml	# added a few endpoints
    │   └── orchestration
    │       ├── __init__.py
    │       ├── alcf_transfer.py	# transfer to ALCF in a module (not currently using this)
    │       ├── config.py	# stayed the same
    │       ├── flows
    │       │   └── bl832
    │       │       ├── ALCF_tomopy_reconstruction.py	# equivalent to move.py
    │       │       ├── config.py	# added a few endpoints
    │       │       └── test
    │       │           ├── alcf_compute_endpoint_test.py	# test compute endpoint
    │       │           ├── alcf_transfer_endpoint_test.py	# test transfer endpoint
    │       │           ├── inputOneSliceOfEach.txt	# used for reconstruction function input
    │       │           ├── reconstruction.py	# copy of reconstruction code on ALCF
    │       │           ├── test_cc_auth.py	# test endpoint read/write for service client
    │       │           └── transfer_test.ipynb 
    │       ├── globus.py	# stayed the same
    │       ├── globus_flows_utils.py	# copied from the notebook, modified "get_specific_flow_client()"
    │       └── globus_tomopy_flow_init.py	# copied globus flow definition from the notebook into a module, but the notebook feels more user-friendly for defining a new flow.

# Updates to `orchestration/flows/bl832/move.py`
The main changes to the production code are the following fuctions:
### [new] `transfer_data_to_alcf()`
Followed structure of `transfer_data_to_nersc()`
### [new] `alcf_tomopy_reconstruction_flow()`

 1. Read data from ALCF
 2. Run reconstruction on ALCF compute endpoint
 3. Save reconstruction to ALCF

**Note**: this function needs to be updated to take as inputs:

 - `function_inputs  =  {"rundir":  "/eagle/IRIBeta/als/sea_shell_test"}` (used as input for `reconstruction.py`)
 - Source path
 - Destination path

### [modified] `process_new_832_file_flow(...,  send_to_alcf=False)`
 Snippet:

       ...
       # Send data from NERSC to ALCF (default is False), process it using Tomopy, and send it back to NERSC
        if not is_export_control and send_to_alcf:            
            # Transfer data from NERSC to ALCF
            transfer_success = transfer_data_to_alcf(fp, config.tc, config.nersc_alsdev, config.alcf_iribeta_cgs)
		    ...
            # Run the Tomopy reconstruction flow
            alcf_tomopy_reconstruction_flow()
		    ...
            # Send reconstructed data to NERSC
            transfer_success = transfer_data_to_nersc(file_path, config.tc, config.alcf_iribeta_cgs, config.nersc_alsdev)
      ...

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1207447610810361
  - https://app.asana.com/0/0/1207595684811250
  - https://app.asana.com/0/0/1207971725336001
  - https://app.asana.com/0/0/1208141006139278
  - https://app.asana.com/0/0/1208141006139276

davramov commented:
# Update

As discussed with Dylan, I removed the ALCF_compute_test directory and copied my useful test scripts to the `/examples` folder. Additionally, instead of merging the ACLF transfer/compute flow into `bl832/move.py,` 
I have added `bl832/ALCF_compute_reconstruction.py` as a separate module.

**Note**:
Data transfer in the following function still needs inputs defined (file source/destination and the required .txt file for reconstruction), which we will determine in Monday's (6/17) meeting with Dula.
```
@flow(name="alcf_tomopy_reconstruction_flow")
def alcf_tomopy_reconstruction_flow():
```

davramov commented:
### June 28 / July 1st Update

I have updated the `ALCF_tomopy_reconstruction.py` script and corresponding files (`reconstruction.py`, `.env`) to account for a file name and folder name when calling the reconstruction Globus Flow. I have also included a specific readme  `README_ALCF_tomopy_reconstruction.md` with an outline of the entire flow, details about folder and file names, step-by-step instructions for configuring the Globus environments, NERSC and ALCF endpoints (transfer and compute), setting up and logging into the confidential client, and running the script.

dylanmcreynolds reviewed (changes requested):
You should remove `examples/Untitled.ipynb`

dylanmcreynolds reviewed (changes requested):
Ultra picky, but can we rename `globus_flows_utils.py` to just `globus_flows.py`?

dylanmcreynolds reviewed (changes requested):
What do you think of moving `README_ALCO_tompy_recon.py` into `/docs`?

Actually, what about a new folder `/orchestration/scripts/` so that we can separate code that runs continuously in flows from code that it run very rarely?

davramov commented:
July 26 Update

**Major changes**
Incorporate `tiff_to_zarr.py` into Globus Flow on ALCF Polaris
File pruning with Prefect worker scheduling
updated `orchestration/flow/bl832/prune.py`
**Other changes:**
`config.yml` and `orchestration/flows/bl832/config.py`
Cleaned up some endpoint names for consistency
`docs/README_alcf832.md` updated to reflect recent progress
**New files:**
`create_deployments_832_alcf.sh`
This shell script builds and deploys pruning code for this workflow.
`examples/Tomopy_for_ALS.ipynb`
This notebook is also in another repository, but this version matches the changes made to the registered flow function and confidential client for this specific workflow.
`examples/tiff_to_zarr.py`
This program reflects the code on Polaris `/eagle/IRIbeta/als/examples/tiff_to_zarr.py` used in the Globus Flow.
**Moved and renamed files:**
`examples/test_cc_auth.py`
`scripts/globus_tomopy_flow_init.py`
Todo: update this script so we can move away from using a Jupyter Notebook to initialize the steps.
Globus helper code
`orchestration/globus/flows.py` (was `globus_flows_utils.py`)
`orchestration/globus/transfer.py` (was `orchestration/globus.py`)
`orchestration/flows/bl832/alcf.py` (was `ALCF_tomopy_reconstruction.py`)
`docs/README_aclf832.md` (was `orchestration/flows/bl832/README_ALCF_tomopy_reconstruction.md`)
**Files with modified imports due to moving and renaming the Globus helper code, but are otherwise unchanged:**
`orchestration/flows/bl7012/config.py`
`orchestration/flows/bl7012/move.py`
`orchestration/flows/bl7012/move_recon.py`
`orchestration/flows/bl832/move.py`
`orchestration/prefect.py`

dylanmcreynolds reviewed (changes requested):
Some thoughts on the newly-added `test_globus_flow.py`. For the most part it is testing data transfers and not flows.  It also copy/pastes some code from `test_globus.py`.

dylanmcreynolds commented on orchestration/_tests/test_globus_flow.py:
```diff
@@ -0,0 +1,154 @@
+    assert result
+
+
+def test_transfer_data_to_nersc():
```
Unit Tests need to be more generic than this. `test_transfer_data_to_alcf` and `test_transfer_data_to_data832` aren't testing anything about nersc and alcf, and should probably be removed.

dylanmcreynolds commented on orchestration/_tests/test_globus_flow.py:
```diff
@@ -0,0 +1,154 @@
+    path = "/path/to/prune"
+    location = "test_location"
+    schedule_days = datetime.timedelta(days=0.01)
+    result = schedule_prune_task(path, location, schedule_days)
```
You're running some code, but maybe not testing it? It seems to me that success here would be tested by using the prefect api to test if the pruning task was succesfully scheduled. This feels like it belongs in a new `test_prefect.py` module.

dylanmcreynolds commented on orchestration/_tests/test_globus_flow.py:
```diff
@@ -0,0 +1,154 @@
+import time
+
+
+class MockTransferClient:
```
These mock classes are copy/pastes from `test_globus.py` and should be removed.

dylanmcreynolds commented:
To get this going, we can pin prefect to 2.7, which is very old and brutal. I created an issue with globus to be less specific on the pin: https://github.com/globus/globus-compute/issues/1624.

dylanmcreynolds commented on orchestration/_tests/test_832.py:
```diff
@@ -1,10 +1,18 @@
+with patch('prefect.blocks.system.Secret.load', return_value=mock_secret):
+    # Import modules that might use Prefect secrets
+    from ..flows.bl832 import move
+    from orchestration.flows.bl832.move import process_new_832_file
```
Interestingly, as it stands, this entire module does nothing and its intended functionality is now in test_globus.py. Let's delete this module.

dylanmcreynolds commented on orchestration/_tests/test_globus_flow.py:
```diff
@@ -0,0 +1,421 @@
+        yield instance
+
+
+class FlowDefinition(BaseModel):
```
It would be interseting to see these models used in the actual code rather than just the tests. Perhaps a future PR?

dylanmcreynolds commented on orchestration/_tests/test_globus_flow.py:
```diff
@@ -0,0 +1,421 @@
+        self.alcf832_scratch = mock_config.alcf832_scratch
+        self.nersc832_alsdev_scratch = mock_config.nersc832_alsdev_scratch
+
+    monkeypatch.setattr(Config832, "__init__", mock_init)
```
This feels unnecessary. You can either choose to use the `MockConfig832` above or monkeypatch the unmocked `Config832`. I vote for the `MockConfig832`

dylanmcreynolds commented on orchestration/_tests/test_globus_flow.py:
```diff
@@ -0,0 +1,421 @@
+            "expires_in": 3600,
+        }
+
+    monkeypatch.setattr(ConfidentialAppAuthClient,
```
This feels like overkill to me. You have a `MockGlobusComputeClient` class defined above. Using it would seem to avoid having to mokeypatch any of the internals of Globus flow clients, like the `ConifdentialAppAuthClient` stuff.

dylanmcreynolds commented on docs/README_alcf832.md:
```diff
@@ -0,0 +1,443 @@
+# 8.3.2 ALCF Globus Flow And Reconstruction Setup
+
+This script facilitates the transfer and processing of data files from the Advanced Light Source (ALS) at Berkeley Lab to the Argonne Leadership Computing Facility (ALCF) and the National Energy Research Scientific Computing Center (NERSC). It includes functions for transferring data using Globus and for performing tomographic reconstruction using Tomopy at ALCF.
```
Now that we moved this into docs, the script name is a little ambigious. For this, we could use a markdown link...  [the init script](../orchestration/scritps/globaus_tomopy_flow_init.py)

dylanmcreynolds reviewed (changes requested):
Looking at the structure of this again, I would like to move script to the root folder and remove the `__init__.py`. It would make them easier to find, and scripts like these should probably not be modules in our API.

What do you think of this structure?
 
|_scripts
   |_alcf
      |_check_globus_compute_status.py
      |_globus_tomopy_flow_init.py
      |_README.md

dylanmcreynolds commented on orchestration/scripts/globus_tomopy_flow_init.py:
```diff
@@ -0,0 +1,141 @@
+from globus_compute_sdk import Client, Executor
+from prefect.blocks.system import Secret
```
Is there a compelling reason for this to be a prefect flow?

dylanmcreynolds commented on scripts/README.md:
```diff
@@ -0,0 +1,544 @@
+# 8.3.2 ALCF Globus Flow And Reconstruction Setup
+
+This script facilitates transferring and processing raw tomography data from Beamline 8.3.2 at Berkeley Lab's Advanced Light Source (ALS) to the Argonne Leadership Computing Facility (ALCF). This script relies on Prefect for workflow orchestration, Globus Transfer for data movement between facilities, and Globus Compute at ALCF for tomographic reconstruction using Tomopy.
```
Maybe clarify which script you're referring to with `This script`?

dylanmcreynolds commented on scripts/README.md:
```diff
@@ -0,0 +1,544 @@
+
+## Globus Compute Flow
+
+The preferred method for scheduling jobs on ALCF is through Globus Compute. This `reconstruction_wrapper()` function contains the code registered with Globus, which runs the reconstruction on ALCF. The commands call two Python programs located in the IRIBeta project folder on Polaris. Here is an example of code that is registered and run in the Globus Compute Flow:
```
Rather than copying code here, maybe just link to the file? Just concerned about this copy getting outdated.

dylanmcreynolds replied:
What file is this referring to?

dylanmcreynolds commented on scripts/init_tiff_to_zarr_globus_flow.py:
```diff
@@ -0,0 +1,103 @@
+    return compute_endpoint_id
+
+
+def reconstruction_wrapper(rundir, h5_file_name, folder_path):
```
Maybe rename to `conversion_wrapper`?

dylanmcreynolds commented on scripts/README.md:
```diff
@@ -0,0 +1,544 @@
+
+The `globus_reconstruction.py` script in this flow is a slightly modified version of Dula Parkinson's tomography reconstruction code.  The modifications update the function's input parameters and set proper file/folder permissions for the reconstructed data.
+
+Location on Polaris: `/eagle/IRIBeta/als/example/globus_reconstruction.py`
```
These files are just placed on polaris? If they're not under version control, perhaps we can put them in the repo under `/script/polaris`?  Same with the tiff_to_zarr.py below

dylanmcreynolds commented on orchestration/_tests/test_globus_flow.py:
```diff
@@ -0,0 +1,339 @@
+    properties: Dict[str, Dict[str, Any]]
+
+
+class CreateFlowRequest(BaseModel):
```
I am a big fan of using pydantic to define models. It's an important tool for run-time validation of the items of dict/json object. It would be cool if similar type-checking were provided by the globus SDK.

However, in this case, the `CreateFlowRequest` and `RunFlowRequest` model and all of its sub-models are doing a lot more work than you need,  being used only once each in your code. And since you control the inputs in your test, you don't really need runtime validation.

I would like to remove these classes and when just provide in in-line dict for the test case where they are used.

dylanmcreynolds commented on orchestration/_tests/test_globus_flow.py:
```diff
@@ -0,0 +1,333 @@
+    States: Dict[str, Dict[str, Any]]
+
+
+class FlowInputSchema(BaseModel):
```
I would like to remove FlowInputSchema, FlowDefinition and the prefect imports.

dylanmcreynolds replied:
The previous comment had a typo. I meant to ask to remove the pydantic imports, which won't be needed when we remove the `BaseModel` sub-classes.
